src/Symfony/Component/Security/Http/Firewall/SwitchUserListener.php

```diff
@@ -162,7 +163,7 @@ private function attemptExitUser(Request $request)
             throw new AuthenticationCredentialsNotFoundException('Could not find original Token object.');
         }
 
-        if (null !== $this->dispatcher) {
+        if (null !== $this->dispatcher && !$original instanceof AnonymousToken) {
```

Comment:
I think it would be better to check if `$original->getUser` is an instance of the `UserInterface` to also support custom tokens that might return something else.